{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_type(barcode):\n",
    "    # example: get \"01\" from TCGA-W5-A22X-01A\n",
    "    # 00-09 tumor; 10-19 normal; 20-29 control\n",
    "    match = re.search(r'-([0-2][0-9])[A-Z]*', barcode)\n",
    "    if match:\n",
    "        code = int(match.group(1))\n",
    "        if code < 10:\n",
    "            return \"TUMOR\"\n",
    "        elif code < 20:\n",
    "            return \"NORMAL\"\n",
    "        elif code < 30:\n",
    "            return \"CONTROL\"\n",
    "        else:\n",
    "            raise SyntaxError(\"Its not a TCGA barcode format.\")\n",
    "\n",
    "def open_vcf(vcf_file):\n",
    "    # Find the line number containing string \"#CHROM\"\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        for line_ind, line in enumerate(f):\n",
    "            if line.startswith('#CHROM'):\n",
    "                break\n",
    "    \n",
    "    # load data and skip headers\n",
    "    df = pd.read_csv(vcf_file, sep='\\t', header=line_ind, dtype=object)\n",
    "    return df\n",
    "\n",
    "def preprocessing_vcf(vcf_file):\n",
    "    vcf_df = open_vcf(vcf_file)\n",
    "    pass_filter = vcf_df.FILTER == \"PASS\"\n",
    "    # print(pass_filter.sum(), pass_filter.count(), pass_ratio)\n",
    "    vcf_df = vcf_df.loc[pass_filter] # keep \"PASS\"\n",
    "    vcf_df = vcf_df.drop(columns=['ID', 'QUAL', 'FILTER'])#, 'FORMAT'])\n",
    "    # keep only chr1-22 or XY\n",
    "    vcf_df = vcf_df[vcf_df['#CHROM'].str.contains(r'^chr(?:[1-9]|1\\d|2[0-2]|X|Y)$')]\n",
    "    vcf_df.columns = ['CHROM'] + list(vcf_df.columns[1:].str.upper())\n",
    "\n",
    "    # rename last 2 columns to tumor/normal\n",
    "    #\n",
    "    vcf_df.index = range(len(vcf_df))\n",
    "    return vcf_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"/home/data/dataset/CHOL_10sample/\")\n",
    "sample_sheet = pd.read_csv(\"./gdc_sample_sheet.2024-02-26.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# sample_sheet = pd.read_csv(\"/home/data/dataset/CHOL_2sample/gdc_sample_sheet_test.2024-02-26.tsv\", sep=\"\\t\")\n",
    "#case_ID\n",
    "#for sample_ID in sample_sheet.:\n",
    "#gb = sample_sheet.groupby(\"Case ID\")\n",
    "#for case_ID in gb.groups:\n",
    "#    print(case_ID)\n",
    "\n",
    "for case_ID in sample_sheet.loc[:, \"Case ID\"]:\n",
    "    muse = preprocessing_vcf(vcf_file=f\"variants_calling/muse/{case_ID}.MuSE.vcf\")\n",
    "    mutect2 = preprocessing_vcf(vcf_file=f\"variants_calling/mutect_only_vcf/{case_ID}_filtered.vcf\")\n",
    "    mutect2 = mutect2.rename(columns={mutect2.columns[-1]: sample_type(mutect2.columns[-1]), \n",
    "                                      mutect2.columns[-2]: sample_type(mutect2.columns[-2])})\n",
    "    varscan_indel = preprocessing_vcf(vcf_file=f\"variants_calling/varscan_GDC/{case_ID}.varscan.indel.Somatic.hc.vcf\")\n",
    "    varscan_snp = preprocessing_vcf(vcf_file=f\"variants_calling/varscan_GDC/{case_ID}.varscan.snp.Somatic.hc.vcf\")\n",
    "    varscan = pd.concat([varscan_indel, varscan_snp], ignore_index=True)\n",
    "    # 合併邏輯"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_info(package_name, callformat):\n",
    "    '''\n",
    "    varscan, GT:GQ:DP:RD:AD:FREQ:DP4, 0/0:.:68:65:2:2.99%:15,50,1,1\n",
    "    mutect2, GT:AD:AF:DP:F1R2:F2R1, 0/1:12,4:0.275:16:8,2:4,2\n",
    "    muse, GT:DP:AD:BQ:SS, 0/1:19:14,5:31,31:2\t\n",
    "    '''\n",
    "    if package_name == \"varscan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TUMOR</th>\n",
       "      <th>NORMAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8967564</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:19:14,5:31,31:2</td>\n",
       "      <td>0/0:26:26,0:31,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16045536</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:24:17,7:28,29:2</td>\n",
       "      <td>0/0:42:42,0:28,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27550441</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:13:8,5:29,31:2</td>\n",
       "      <td>0/0:23:23,0:29,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43981297</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:30:21,9:32,30:2</td>\n",
       "      <td>0/0:34:34,0:33,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61762741</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:54:38,16:29,31:2</td>\n",
       "      <td>0/0:74:74,0:29,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>chr22</td>\n",
       "      <td>37206698</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:75:65,10:30,29:2</td>\n",
       "      <td>0/0:81:81,0:30,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>chr22</td>\n",
       "      <td>41589133</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:39:31,8:30,28:2</td>\n",
       "      <td>0/0:50:50,0:30,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>chrX</td>\n",
       "      <td>106903064</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:27:17,10:34,29:2</td>\n",
       "      <td>0/0:19:19,0:34,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>chrX</td>\n",
       "      <td>120604680</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:13:5,8:35,28:2</td>\n",
       "      <td>0/0:14:14,0:33,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26671288</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:15:12,3:31,34:2</td>\n",
       "      <td>0/0:25:25,0:31,0:.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS REF ALT     INFO          FORMAT                 TUMOR  \\\n",
       "0     chr1    8967564   A   G  SOMATIC  GT:DP:AD:BQ:SS   0/1:19:14,5:31,31:2   \n",
       "1     chr1   16045536   G   A  SOMATIC  GT:DP:AD:BQ:SS   0/1:24:17,7:28,29:2   \n",
       "2     chr1   27550441   C   T  SOMATIC  GT:DP:AD:BQ:SS    0/1:13:8,5:29,31:2   \n",
       "3     chr1   43981297   G   T  SOMATIC  GT:DP:AD:BQ:SS   0/1:30:21,9:32,30:2   \n",
       "4     chr1   61762741   T   C  SOMATIC  GT:DP:AD:BQ:SS  0/1:54:38,16:29,31:2   \n",
       "..     ...        ...  ..  ..      ...             ...                   ...   \n",
       "115  chr22   37206698   T   A  SOMATIC  GT:DP:AD:BQ:SS  0/1:75:65,10:30,29:2   \n",
       "116  chr22   41589133   A   T  SOMATIC  GT:DP:AD:BQ:SS   0/1:39:31,8:30,28:2   \n",
       "117   chrX  106903064   C   A  SOMATIC  GT:DP:AD:BQ:SS  0/1:27:17,10:34,29:2   \n",
       "118   chrX  120604680   G   T  SOMATIC  GT:DP:AD:BQ:SS    0/1:13:5,8:35,28:2   \n",
       "119   chrY   26671288   A   G  SOMATIC  GT:DP:AD:BQ:SS   0/1:15:12,3:31,34:2   \n",
       "\n",
       "                 NORMAL  \n",
       "0    0/0:26:26,0:31,0:.  \n",
       "1    0/0:42:42,0:28,0:.  \n",
       "2    0/0:23:23,0:29,0:.  \n",
       "3    0/0:34:34,0:33,0:.  \n",
       "4    0/0:74:74,0:29,0:.  \n",
       "..                  ...  \n",
       "115  0/0:81:81,0:30,0:.  \n",
       "116  0/0:50:50,0:30,0:.  \n",
       "117  0/0:19:19,0:34,0:.  \n",
       "118  0/0:14:14,0:33,0:.  \n",
       "119  0/0:25:25,0:31,0:.  \n",
       "\n",
       "[120 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "format_str = \"GT:GQ:DP:RD:AD:FREQ:DP4\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'DP*' is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m muse_format \u001b[38;5;241m=\u001b[39m muse\u001b[38;5;241m.\u001b[39mFORMAT[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m DP_index \u001b[38;5;241m=\u001b[39m \u001b[43mmuse_format\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mDP*\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m      5\u001b[0m     AF_index \u001b[38;5;241m=\u001b[39m muse_format\u001b[38;5;241m.\u001b[39mindex(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAF\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: 'DP*' is not in list"
     ]
    }
   ],
   "source": [
    "muse_format = muse.FORMAT[0].split(\":\")\n",
    "\n",
    "DP_index = muse_format.index(\"DP*\")\n",
    "try:\n",
    "    AF_index = muse_format.index(\"AF\")\n",
    "except ValueError:\n",
    "    AD_index = muse_format.index(\"AD\")\n",
    "muse_format.index(\"DP\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "muse.columns = ['CHROM'] + list(muse.columns[1:].str.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    TCGA-W5-AA2X-01A\n",
       "1    TCGA-W5-AA2X-11A\n",
       "Name: Sample ID, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb.get_group(\"TCGA-W5-AA2X\").loc[:, \"Sample ID\"].apply()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('tumor', 'normal')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2.columns[-2:] = sample_type(mutect2.columns[-2]), sample_type(mutect2.columns[-1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ID = \"TCGA-W5-AA2X-01A\"\t\n",
    "mutect2.columns = mutect2.columns[:-2] + mutect2.columns[-2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TCGA-W5-AA2X-01A</th>\n",
       "      <th>TCGA-W5-AA2X-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1390825</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CONTQ=53;DP=171;ECNT=1;GERMQ=402;MBQ=36,30;MFR...</td>\n",
       "      <td>0/1:49,3:0.074:52:21,0:28,3</td>\n",
       "      <td>0/0:113,0:8.570e-03:113:40,0:73,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2149147</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=64;DP=12;ECNT=1;GERMQ=22;MBQ=35,34;MFRL=...</td>\n",
       "      <td>0/1:1,2:0.604:3:0,0:1,2</td>\n",
       "      <td>0/0:8,0:0.100:8:3,0:5,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2156164</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=51;DP=19;ECNT=1;GERMQ=30;MBQ=35,34;MFRL=...</td>\n",
       "      <td>0/1:6,2:0.300:8:3,2:3,0</td>\n",
       "      <td>0/0:10,0:0.081:10:2,0:8,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3635943</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=65;DP=94;ECNT=1;GERMQ=213;MBQ=37,32;MFRL...</td>\n",
       "      <td>0/1:24,3:0.137:27:11,0:13,3</td>\n",
       "      <td>0/0:64,0:0.015:64:18,0:46,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11957858</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CONTQ=93;DP=153;ECNT=1;GERMQ=334;MBQ=33,30;MFR...</td>\n",
       "      <td>0/1:26,13:0.341:39:10,3:16,10</td>\n",
       "      <td>0/0:111,0:8.789e-03:111:46,0:65,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7660</th>\n",
       "      <td>chrUn_JTFH01001430v1_decoy</td>\n",
       "      <td>222</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>CONTQ=74;DP=38;ECNT=2;GERMQ=71;MBQ=33,35;MFRL=...</td>\n",
       "      <td>0/1:13,3:0.221:16:6,1:7,2</td>\n",
       "      <td>0/0:20,0:0.045:20:8,0:12,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7681</th>\n",
       "      <td>chrUn_JTFH01001478v1_decoy</td>\n",
       "      <td>500</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>CONTQ=43;DP=33;ECNT=1;GERMQ=70;MBQ=36,32;MFRL=...</td>\n",
       "      <td>0/1:13,2:0.176:15:7,0:6,2</td>\n",
       "      <td>0/0:18,0:0.050:18:8,0:10,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7692</th>\n",
       "      <td>chrUn_JTFH01001553v1_decoy</td>\n",
       "      <td>782</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CONTQ=70;DP=13;ECNT=1;GERMQ=33;MBQ=33,36;MFRL=...</td>\n",
       "      <td>0/1:0,2:0.750:2:0,1:0,1</td>\n",
       "      <td>0/0:10,0:0.081:10:6,0:4,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7705</th>\n",
       "      <td>chrUn_JTFH01001847v1_decoy</td>\n",
       "      <td>1652</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=87;DP=16;ECNT=1;GERMQ=22;MBQ=37,32;MFRL=...</td>\n",
       "      <td>0/1:5,3:0.400:8:3,2:2,1</td>\n",
       "      <td>0/0:8,0:0.100:8:2,0:6,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7720</th>\n",
       "      <td>chrUn_JTFH01001998v1_decoy</td>\n",
       "      <td>1799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CONTQ=77;DP=27;ECNT=1;GERMQ=43;MBQ=33,35;MFRL=...</td>\n",
       "      <td>0/1:11,3:0.249:14:5,1:6,2</td>\n",
       "      <td>0/0:12,0:0.071:12:10,0:2,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>364 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          #CHROM       POS REF ALT  \\\n",
       "6                           chr1   1390825   G   A   \n",
       "12                          chr1   2149147   G   T   \n",
       "13                          chr1   2156164   G   T   \n",
       "17                          chr1   3635943   C   T   \n",
       "37                          chr1  11957858   T   C   \n",
       "...                          ...       ...  ..  ..   \n",
       "7660  chrUn_JTFH01001430v1_decoy       222   T   G   \n",
       "7681  chrUn_JTFH01001478v1_decoy       500   G   C   \n",
       "7692  chrUn_JTFH01001553v1_decoy       782   T   C   \n",
       "7705  chrUn_JTFH01001847v1_decoy      1652   C   T   \n",
       "7720  chrUn_JTFH01001998v1_decoy      1799   A   G   \n",
       "\n",
       "                                                   INFO  \\\n",
       "6     CONTQ=53;DP=171;ECNT=1;GERMQ=402;MBQ=36,30;MFR...   \n",
       "12    CONTQ=64;DP=12;ECNT=1;GERMQ=22;MBQ=35,34;MFRL=...   \n",
       "13    CONTQ=51;DP=19;ECNT=1;GERMQ=30;MBQ=35,34;MFRL=...   \n",
       "17    CONTQ=65;DP=94;ECNT=1;GERMQ=213;MBQ=37,32;MFRL...   \n",
       "37    CONTQ=93;DP=153;ECNT=1;GERMQ=334;MBQ=33,30;MFR...   \n",
       "...                                                 ...   \n",
       "7660  CONTQ=74;DP=38;ECNT=2;GERMQ=71;MBQ=33,35;MFRL=...   \n",
       "7681  CONTQ=43;DP=33;ECNT=1;GERMQ=70;MBQ=36,32;MFRL=...   \n",
       "7692  CONTQ=70;DP=13;ECNT=1;GERMQ=33;MBQ=33,36;MFRL=...   \n",
       "7705  CONTQ=87;DP=16;ECNT=1;GERMQ=22;MBQ=37,32;MFRL=...   \n",
       "7720  CONTQ=77;DP=27;ECNT=1;GERMQ=43;MBQ=33,35;MFRL=...   \n",
       "\n",
       "                   TCGA-W5-AA2X-01A                   TCGA-W5-AA2X-11A  \n",
       "6       0/1:49,3:0.074:52:21,0:28,3  0/0:113,0:8.570e-03:113:40,0:73,0  \n",
       "12          0/1:1,2:0.604:3:0,0:1,2            0/0:8,0:0.100:8:3,0:5,0  \n",
       "13          0/1:6,2:0.300:8:3,2:3,0          0/0:10,0:0.081:10:2,0:8,0  \n",
       "17      0/1:24,3:0.137:27:11,0:13,3        0/0:64,0:0.015:64:18,0:46,0  \n",
       "37    0/1:26,13:0.341:39:10,3:16,10  0/0:111,0:8.789e-03:111:46,0:65,0  \n",
       "...                             ...                                ...  \n",
       "7660      0/1:13,3:0.221:16:6,1:7,2         0/0:20,0:0.045:20:8,0:12,0  \n",
       "7681      0/1:13,2:0.176:15:7,0:6,2         0/0:18,0:0.050:18:8,0:10,0  \n",
       "7692        0/1:0,2:0.750:2:0,1:0,1          0/0:10,0:0.081:10:6,0:4,0  \n",
       "7705        0/1:5,3:0.400:8:3,2:2,1            0/0:8,0:0.100:8:2,0:6,0  \n",
       "7720      0/1:11,3:0.249:14:5,1:6,2         0/0:12,0:0.071:12:10,0:2,0  \n",
       "\n",
       "[364 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2, pass_ratio = preprocessing_vcf(\"mutect2/TCGA-W5-AA2X_filtered.vcf\")\n",
    "mutect2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0471441523118767"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2 = open_vcf(\"mutect2/TCGA-W5-AA2X_filtered.vcf\")\n",
    "pass_filter = mutect2.FILTER == \"PASS\"\n",
    "pass_ratio = pass_filter.sum()/pass_filter.count()\n",
    "print(pass_filter.sum(), pass_filter.count(), pass_ratio)\n",
    "\n",
    "mutect2 = mutect2.loc[pass_filter] # keep \"PASS\"\n",
    "mutect2 = mutect2.drop(columns=['ID', 'QUAL', 'FILTER', 'FORMAT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'normal'"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_type(mutect2.columns[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "muse = open_vcf(\"muse/TCGA-W5-AA2X.MuSE.vcf\")\n",
    "pass_filter = muse.FILTER == \"PASS\"\n",
    "muse = muse.loc[pass_filter] # keep \"PASS\"\n",
    "muse = muse.drop(columns=['ID', 'QUAL', 'FILTER'])\n",
    "# drop CHROM not in 1-22 or XY\n",
    "muse = muse[muse['#CHROM'].str.contains(r'^chr(?:[1-9]|1\\d|2[0-2]|X|Y)$')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TUMOR</th>\n",
       "      <th>NORMAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2156164</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:8:6,2:35,34:2</td>\n",
       "      <td>0/0:10:10,0:35,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11957858</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:50:33,17:31,28:2</td>\n",
       "      <td>0/0:144:140,3:30,2:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>26773569</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:14:6,8:33,36:2</td>\n",
       "      <td>0/0:66:66,0:32,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33772581</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:11:5,6:37,27:2</td>\n",
       "      <td>0/0:14:14,0:36,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr1</td>\n",
       "      <td>45530932</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:10:7,3:37,34:2</td>\n",
       "      <td>0/0:20:20,0:34,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1147</th>\n",
       "      <td>chrX</td>\n",
       "      <td>55221659</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:30:7,23:32,37:2</td>\n",
       "      <td>0/0:47:47,0:31,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>chrX</td>\n",
       "      <td>72712853</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>1/1:28:4,24:31,33:2</td>\n",
       "      <td>0/0:37:37,0:30,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>chrX</td>\n",
       "      <td>103724114</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:15:5,10:33,32:2</td>\n",
       "      <td>0/0:35:35,0:33,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12859165</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>0/1:65:56,9:31,30:2</td>\n",
       "      <td>0/0:121:121,0:30,0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21383847</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>GT:DP:AD:BQ:SS</td>\n",
       "      <td>1/1:28:5,23:32,32:2</td>\n",
       "      <td>0/0:50:50,0:30,0:.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>162 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS REF ALT     INFO          FORMAT                 TUMOR  \\\n",
       "3      chr1    2156164   G   T  SOMATIC  GT:DP:AD:BQ:SS     0/1:8:6,2:35,34:2   \n",
       "9      chr1   11957858   T   C  SOMATIC  GT:DP:AD:BQ:SS  0/1:50:33,17:31,28:2   \n",
       "24     chr1   26773569   A   G  SOMATIC  GT:DP:AD:BQ:SS    0/1:14:6,8:33,36:2   \n",
       "33     chr1   33772581   G   A  SOMATIC  GT:DP:AD:BQ:SS    0/1:11:5,6:37,27:2   \n",
       "40     chr1   45530932   C   A  SOMATIC  GT:DP:AD:BQ:SS    0/1:10:7,3:37,34:2   \n",
       "...     ...        ...  ..  ..      ...             ...                   ...   \n",
       "1147   chrX   55221659   A   C  SOMATIC  GT:DP:AD:BQ:SS   0/1:30:7,23:32,37:2   \n",
       "1154   chrX   72712853   A   G  SOMATIC  GT:DP:AD:BQ:SS   1/1:28:4,24:31,33:2   \n",
       "1156   chrX  103724114   C   T  SOMATIC  GT:DP:AD:BQ:SS   0/1:15:5,10:33,32:2   \n",
       "1177   chrY   12859165   A   T  SOMATIC  GT:DP:AD:BQ:SS   0/1:65:56,9:31,30:2   \n",
       "1179   chrY   21383847   C   T  SOMATIC  GT:DP:AD:BQ:SS   1/1:28:5,23:32,32:2   \n",
       "\n",
       "                    NORMAL  \n",
       "3       0/0:10:10,0:35,0:.  \n",
       "9     0/0:144:140,3:30,2:.  \n",
       "24      0/0:66:66,0:32,0:.  \n",
       "33      0/0:14:14,0:36,0:.  \n",
       "40      0/0:20:20,0:34,0:.  \n",
       "...                    ...  \n",
       "1147    0/0:47:47,0:31,0:.  \n",
       "1154    0/0:37:37,0:30,0:.  \n",
       "1156    0/0:35:35,0:33,0:.  \n",
       "1177  0/0:121:121,0:30,0:.  \n",
       "1179    0/0:50:50,0:30,0:.  \n",
       "\n",
       "[162 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "varscan_snp = open_vcf(\"muse/TCGA-W5-AA2X.MuSE.vcf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "#mutect2.columns[-1] = sample_type(mutect2.columns[-1])\n",
    "#mutect2.columns[-2] = sample_type(mutect2.columns[-2])\n",
    "\n",
    "# drop CHROM not in 1-22 or XY\n",
    "mutect2 = mutect2[mutect2['#CHROM'].str.contains(r'^chr(?:[1-9]|1\\d|2[0-2]|X|Y)$')]\n",
    "\n",
    "# explode multihit - 一個REF對到多個ALT(少見)\n",
    "multi_hit_filter = mutect2['ALT'].str.contains(',')\n",
    "multi_hit_split = mutect2.loc[multi_hit_filter, 'ALT'].str.split(',')\n",
    "mutect2.loc[multi_hit_filter, 'ALT'] = multi_hit_split\n",
    "mutect2 = mutect2.explode('ALT')\n",
    "mutect2 = mutect2.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CONTQ=53;DP=171;ECNT=1;GERMQ=402;MBQ=36,30;MFRL=215,174;MMQ=60,60;MPOS=29;NALOD=2.06;NLOD=34.00;POPAF=6.00;SAAF=0.030,0.051,0.058;SAPP=0.028,5.545e-03,0.967;TLOD=5.48'"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2.INFO[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3605354/702046631.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mutect2_copy = mutect2[multi_hit_filter].copy()\n"
     ]
    },
    {
     "ename": "IndexingError",
     "evalue": "Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexingError\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[92], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m mutect2_copy \u001b[38;5;241m=\u001b[39m \u001b[43mmutect2\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmulti_hit_filter\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m      2\u001b[0m mutect2_copy\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/frame.py:4081\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4079\u001b[0m \u001b[38;5;66;03m# Do we have a (boolean) 1d indexer?\u001b[39;00m\n\u001b[1;32m   4080\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m com\u001b[38;5;241m.\u001b[39mis_bool_indexer(key):\n\u001b[0;32m-> 4081\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_bool_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4083\u001b[0m \u001b[38;5;66;03m# We are left with two options: a single key, and a collection of keys,\u001b[39;00m\n\u001b[1;32m   4084\u001b[0m \u001b[38;5;66;03m# We interpret tuples as collections only for non-MultiIndex\u001b[39;00m\n\u001b[1;32m   4085\u001b[0m is_single_key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mtuple\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_list_like(key)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/frame.py:4137\u001b[0m, in \u001b[0;36mDataFrame._getitem_bool_array\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4131\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   4132\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mItem wrong length \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(key)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m instead of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   4133\u001b[0m     )\n\u001b[1;32m   4135\u001b[0m \u001b[38;5;66;03m# check_bool_indexer will throw exception if Series key cannot\u001b[39;00m\n\u001b[1;32m   4136\u001b[0m \u001b[38;5;66;03m# be reindexed to match DataFrame rows\u001b[39;00m\n\u001b[0;32m-> 4137\u001b[0m key \u001b[38;5;241m=\u001b[39m \u001b[43mcheck_bool_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4139\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key\u001b[38;5;241m.\u001b[39mall():\n\u001b[1;32m   4140\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexing.py:2662\u001b[0m, in \u001b[0;36mcheck_bool_indexer\u001b[0;34m(index, key)\u001b[0m\n\u001b[1;32m   2660\u001b[0m indexer \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39mget_indexer_for(index)\n\u001b[1;32m   2661\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01min\u001b[39;00m indexer:\n\u001b[0;32m-> 2662\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m IndexingError(\n\u001b[1;32m   2663\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnalignable boolean Series provided as \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2664\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mindexer (index of the boolean Series and of \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2665\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mthe indexed object do not match).\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2666\u001b[0m     )\n\u001b[1;32m   2668\u001b[0m result \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   2670\u001b[0m \u001b[38;5;66;03m# fall through for boolean\u001b[39;00m\n",
      "\u001b[0;31mIndexingError\u001b[0m: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match)."
     ]
    }
   ],
   "source": [
    "mutect2_copy = mutect2[multi_hit_filter].copy()\n",
    "mutect2_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TCGA-W5-AA2X-01A</th>\n",
       "      <th>TCGA-W5-AA2X-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>C</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>CAT</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>A</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>AT</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TAAA</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>GAA</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS    REF   ALT  \\\n",
       "1506   chr3  167683840  CATAT     C   \n",
       "1506   chr3  167683840  CATAT   CAT   \n",
       "2306   chr6   36603076    ATT     A   \n",
       "2306   chr6   36603076    ATT    AT   \n",
       "3639  chr10    5458438    TAA     T   \n",
       "3639  chr10    5458438    TAA  TAAA   \n",
       "4820  chr14   33801862   GAAA     G   \n",
       "4820  chr14   33801862   GAAA   GAA   \n",
       "\n",
       "                                                   INFO  \\\n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "\n",
       "                     FORMAT                          TCGA-W5-AA2X-01A  \\\n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "\n",
       "                             TCGA-W5-AA2X-11A  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2_copy = mutect2[multi_hit_filter].copy()\n",
    "mutect2_copy[\"ALT\"] = mutect2_copy[\"ALT\"].str.split(',')\n",
    "mutect2_copy.explode('ALT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Cannot mask with non-boolean array containing NA / NaN values",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[56], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m multi_hit_filter \u001b[38;5;241m=\u001b[39m mutect2[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mALT\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mcontains(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;66;03m#.str.split(',')\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m multi_hit_split \u001b[38;5;241m=\u001b[39m \u001b[43mmutect2\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmulti_hit_filter\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mALT\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m mutect2\u001b[38;5;241m.\u001b[39mloc[multi_hit_filter, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mALT\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m multi_hit_split\n\u001b[1;32m      4\u001b[0m mutect2 \u001b[38;5;241m=\u001b[39m mutect2\u001b[38;5;241m.\u001b[39mexplode(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mALT\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexing.py:1184\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1182\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_scalar_access(key):\n\u001b[1;32m   1183\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_value(\u001b[38;5;241m*\u001b[39mkey, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n\u001b[0;32m-> 1184\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_tuple\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1185\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1186\u001b[0m     \u001b[38;5;66;03m# we by definition only have the 0th axis\u001b[39;00m\n\u001b[1;32m   1187\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexing.py:1368\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1366\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m suppress(IndexingError):\n\u001b[1;32m   1367\u001b[0m     tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_expand_ellipsis(tup)\n\u001b[0;32m-> 1368\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_lowerdim\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtup\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1370\u001b[0m \u001b[38;5;66;03m# no multi-index, so validate all of the indexers\u001b[39;00m\n\u001b[1;32m   1371\u001b[0m tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_tuple_indexer(tup)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexing.py:1089\u001b[0m, in \u001b[0;36m_LocationIndexer._getitem_lowerdim\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1087\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m section\n\u001b[1;32m   1088\u001b[0m         \u001b[38;5;66;03m# This is an elided recursive call to iloc/loc\u001b[39;00m\n\u001b[0;32m-> 1089\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[43mnew_key\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m   1091\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m IndexingError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnot applicable\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexing.py:1191\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1189\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[1;32m   1190\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_deprecated_callable_usage(key, maybe_callable)\n\u001b[0;32m-> 1191\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexing.py:1412\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1410\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[1;32m   1411\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_slice_axis(key, axis\u001b[38;5;241m=\u001b[39maxis)\n\u001b[0;32m-> 1412\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[43mcom\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_bool_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m   1413\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getbool_axis(key, axis\u001b[38;5;241m=\u001b[39maxis)\n\u001b[1;32m   1414\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_list_like_indexer(key):\n\u001b[1;32m   1415\u001b[0m     \u001b[38;5;66;03m# an iterable multi-selection\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/common.py:136\u001b[0m, in \u001b[0;36mis_bool_indexer\u001b[0;34m(key)\u001b[0m\n\u001b[1;32m    132\u001b[0m     na_msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot mask with non-boolean array containing NA / NaN values\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    133\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mis_bool_array(key_array, skipna\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[1;32m    134\u001b[0m         \u001b[38;5;66;03m# Don't raise on e.g. [\"A\", \"B\", np.nan], see\u001b[39;00m\n\u001b[1;32m    135\u001b[0m         \u001b[38;5;66;03m#  test_loc_getitem_list_of_labels_categoricalindex_with_na\u001b[39;00m\n\u001b[0;32m--> 136\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(na_msg)\n\u001b[1;32m    137\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    138\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot mask with non-boolean array containing NA / NaN values"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TCGA-W5-AA2X-01A</th>\n",
       "      <th>TCGA-W5-AA2X-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>C,CAT</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>A,AT</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T,TAAA</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>G,GAA</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS    REF     ALT  \\\n",
       "1506   chr3  167683840  CATAT   C,CAT   \n",
       "2306   chr6   36603076    ATT    A,AT   \n",
       "3639  chr10    5458438    TAA  T,TAAA   \n",
       "4820  chr14   33801862   GAAA   G,GAA   \n",
       "\n",
       "                                                   INFO  \\\n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "\n",
       "                     FORMAT                          TCGA-W5-AA2X-01A  \\\n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "\n",
       "                             TCGA-W5-AA2X-11A  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_hit_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1506     [C, CAT]\n",
       "2306      [A, AT]\n",
       "3639    [T, TAAA]\n",
       "4820     [G, GAA]\n",
       "Name: ALT, dtype: object"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'Series'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[44], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m multi_hit_split \u001b[38;5;241m=\u001b[39m mutect2[multi_hit_filter]\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m      3\u001b[0m mutect2 \u001b[38;5;241m=\u001b[39m mutect2[\u001b[38;5;241m~\u001b[39mmulti_hit_filter] \u001b[38;5;66;03m# drop multi hit\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[43mmutect2\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmulti_hit_filter\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mALT\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m=\u001b[39m mutect2_split[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mALT\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m mutect2_split\u001b[38;5;241m.\u001b[39mexplode(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mALT\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/frame.py:4299\u001b[0m, in \u001b[0;36mDataFrame.__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   4296\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_setitem_array([key], value)\n\u001b[1;32m   4297\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   4298\u001b[0m     \u001b[38;5;66;03m# set column\u001b[39;00m\n\u001b[0;32m-> 4299\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_set_item\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/frame.py:4515\u001b[0m, in \u001b[0;36mDataFrame._set_item\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   4503\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4504\u001b[0m \u001b[38;5;124;03mAdd series to DataFrame in specified column.\u001b[39;00m\n\u001b[1;32m   4505\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4510\u001b[0m \u001b[38;5;124;03mensure homogeneity.\u001b[39;00m\n\u001b[1;32m   4511\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4512\u001b[0m value, refs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sanitize_column(value)\n\u001b[1;32m   4514\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[0;32m-> 4515\u001b[0m     \u001b[43mkey\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\n\u001b[1;32m   4516\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m value\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   4517\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(value\u001b[38;5;241m.\u001b[39mdtype, ExtensionDtype)\n\u001b[1;32m   4518\u001b[0m ):\n\u001b[1;32m   4519\u001b[0m     \u001b[38;5;66;03m# broadcast across multiple columns if necessary\u001b[39;00m\n\u001b[1;32m   4520\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mis_unique \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns, MultiIndex):\n\u001b[1;32m   4521\u001b[0m         existing_piece \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m[key]\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.12/site-packages/pandas/core/indexes/base.py:5358\u001b[0m, in \u001b[0;36mIndex.__contains__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   5323\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__contains__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key: Any) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n\u001b[1;32m   5324\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   5325\u001b[0m \u001b[38;5;124;03m    Return a boolean indicating whether the provided key is in the index.\u001b[39;00m\n\u001b[1;32m   5326\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5356\u001b[0m \u001b[38;5;124;03m    False\u001b[39;00m\n\u001b[1;32m   5357\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 5358\u001b[0m     \u001b[38;5;28;43mhash\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5359\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   5360\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\n",
      "\u001b[0;31mTypeError\u001b[0m: unhashable type: 'Series'"
     ]
    }
   ],
   "source": [
    "multi_hit_filter = mutect2['ALT'].str.contains(',')\n",
    "multi_hit_split = mutect2[multi_hit_filter].copy()\n",
    "\n",
    "mutect2[multi_hit_filter, 'ALT'] = mutect2_split['ALT'].str.split(',')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TCGA-W5-AA2X-01A</th>\n",
       "      <th>TCGA-W5-AA2X-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>C</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>CAT</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>A</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>AT</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TAAA</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>GAA</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS    REF   ALT  \\\n",
       "1506   chr3  167683840  CATAT     C   \n",
       "1506   chr3  167683840  CATAT   CAT   \n",
       "2306   chr6   36603076    ATT     A   \n",
       "2306   chr6   36603076    ATT    AT   \n",
       "3639  chr10    5458438    TAA     T   \n",
       "3639  chr10    5458438    TAA  TAAA   \n",
       "4820  chr14   33801862   GAAA     G   \n",
       "4820  chr14   33801862   GAAA   GAA   \n",
       "\n",
       "                                                   INFO  \\\n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "\n",
       "                     FORMAT                          TCGA-W5-AA2X-01A  \\\n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "\n",
       "                             TCGA-W5-AA2X-11A  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2_split['ALT'].str.split(',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TCGA-W5-AA2X-01A</th>\n",
       "      <th>TCGA-W5-AA2X-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>C,CAT</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>A,AT</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T,TAAA</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>G,GAA</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS    REF     ALT  \\\n",
       "1506   chr3  167683840  CATAT   C,CAT   \n",
       "2306   chr6   36603076    ATT    A,AT   \n",
       "3639  chr10    5458438    TAA  T,TAAA   \n",
       "4820  chr14   33801862   GAAA   G,GAA   \n",
       "\n",
       "                                                   INFO  \\\n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "\n",
       "                     FORMAT                          TCGA-W5-AA2X-01A  \\\n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "\n",
       "                             TCGA-W5-AA2X-11A  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " mutect2[mutect2['ALT'].str.contains(',')].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>TCGA-W5-AA2X-01A</th>\n",
       "      <th>TCGA-W5-AA2X-11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>C</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1506</th>\n",
       "      <td>chr3</td>\n",
       "      <td>167683840</td>\n",
       "      <td>CATAT</td>\n",
       "      <td>CAT</td>\n",
       "      <td>CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1</td>\n",
       "      <td>0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>A</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36603076</td>\n",
       "      <td>ATT</td>\n",
       "      <td>AT</td>\n",
       "      <td>CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2</td>\n",
       "      <td>0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>T</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>chr10</td>\n",
       "      <td>5458438</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TAAA</td>\n",
       "      <td>CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4</td>\n",
       "      <td>0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4820</th>\n",
       "      <td>chr14</td>\n",
       "      <td>33801862</td>\n",
       "      <td>GAAA</td>\n",
       "      <td>GAA</td>\n",
       "      <td>CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1</td>\n",
       "      <td>0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2</td>\n",
       "      <td>0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS    REF   ALT  \\\n",
       "1506   chr3  167683840  CATAT     C   \n",
       "1506   chr3  167683840  CATAT   CAT   \n",
       "2306   chr6   36603076    ATT     A   \n",
       "2306   chr6   36603076    ATT    AT   \n",
       "3639  chr10    5458438    TAA     T   \n",
       "3639  chr10    5458438    TAA  TAAA   \n",
       "4820  chr14   33801862   GAAA     G   \n",
       "4820  chr14   33801862   GAAA   GAA   \n",
       "\n",
       "                                                   INFO  \\\n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "1506  CONTQ=43;DP=64;ECNT=1;GERMQ=56,0;MBQ=37,35,38;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "2306  CONTQ=25;DP=47;ECNT=1;GERMQ=14,0;MBQ=28,31,24;...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "3639  CONTQ=93;DP=94;ECNT=1;GERMQ=2,104;MBQ=33,33,35...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "4820  CONTQ=73;DP=73;ECNT=1;GERMQ=93,0;MBQ=35,37,36;...   \n",
       "\n",
       "                     FORMAT                          TCGA-W5-AA2X-01A  \\\n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "1506  GT:AD:AF:DP:F1R2:F2R1  0/1/2:13,3,2:0.181,0.145:18:3,1,1:10,2,1   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "2306  GT:AD:AF:DP:F1R2:F2R1   0/1/2:11,3,3:0.189,0.201:17:4,1,1:7,2,2   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "3639  GT:AD:AF:DP:F1R2:F2R1   0/1/2:10,2,5:0.126,0.281:17:3,0,1:7,2,4   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "4820  GT:AD:AF:DP:F1R2:F2R1   0/1/2:12,2,3:0.148,0.198:17:3,1,1:9,1,2   \n",
       "\n",
       "                             TCGA-W5-AA2X-11A  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "1506  0/0:26,0,3:0.027,0.123:29:13,0,1:13,0,2  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "2306    0/0:14,0,1:0.047,0.105:15:5,0,1:9,0,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "3639  0/0:42,3,1:0.075,0.031:46:11,0,1:31,3,0  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  \n",
       "4820  0/0:32,0,8:0.021,0.201:40:13,0,5:19,0,3  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutect2_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
